### PR TITLE
PTB LM example now has far better perplexity for both large (72.30) and small (87.17) models

### PR DESCRIPTION
**Pull request by Smerity**
This is as close to SotA as I can get from a standard LSTM without a form of recurrent dropout (see: https://github.com/pytorch/pytorch/issues/561).

Parameters are similar to [Recurrent Neural Network Regularization (Zaremba et al. 2014)](https://arxiv.org/pdf/1409.2329.pdf) and include the one line tweak from [Tying Word Vectors and Word Classifiers: A Loss Framework for Language Modeling (Inan et al. 2016)](https://arxiv.org/pdf/1611.01462.pdf) that substantially improves perplexity whilst minimizing the number of parameters.

This also serves as a better learning example now as well, given it includes dropout (including the use of `model.train()` and `model.eval()`) and also saves / reloads the model with best validation loss.

### CODE REVIEW

**Comment by soumith** [comment]
this is great, and looks good. Thanks!

**Comment by HolyLow** [comment]
As discussed in [https://github.com/pytorch/examples/issues/415](url), many (including myself) fail to reproduce the claimed perplexity as reported in the README. I really want to know what I've done wrong. @soumith @Smerity

**Comment by HolyLow** [comment]
It seems from [https://github.com/pytorch/examples/issues/281](url) that the perplexity in the README is based on PTB dataset while now the default provided dataset is WT2. Maybe the README should be updated.

**Comment by soumith** [comment]
@HolyLow you are right, the README needs to be updated with baselines on WT2. For now, I am removing the perplexity values in the README
